### PR TITLE
Apache Avro parser plugin

### LINKED ISSUE #1630: Support Avro format for kafka producer and consumer?
is there any plan to support Avro format for Kafka producer and consumer and also other queues?

Avro Schema for Influxdb format:

```
{
          "namespace": "telegraf.avro",
          "type": "record",
          "name": "telegraf",
          "fields": [
            {
            "name": "timestamp",
            "type": {"type":"long", "logicalType":"timestamp-millis"}
            },
            {
            "name": "name",
            "type": "string"
            },
            {
            "name": "tags",
            "type": {"type" : "map", "values" : "string"}
            },
            {
            "name": "fields",
            "type": {"type" : "map", "values" : "double"},
            "type": {"type" : "map", "values" : "float"},
            "type": {"type" : "map", "values" : "boolean"},
            "type": {"type" : "map", "values" : "int"},
            "type": {"type" : "map", "values" : "long"},
            "type": {"type" : "map", "values" : "bytes"},
            "type": {"type" : "map", "values" : "string"},
            "type": {"type" : "map", "values" : "null"}
            }
          ]
        }
}
```

### PR DESCRIPTION
**Pull request by emanuele-falzone**
Hi 😃

This PR adds the ability to parse Apache Avro encoded messages.
This plugin is inspired from the use of Apache Avro with Kafka and [Confluent Schema Registry](https://docs.confluent.io/current/schema-registry/index.html).
Issue [#1630](https://github.com/influxdata/telegraf/issues/1630) discusses the need of such parser.

This PR has
- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

Closes #1630

### CODE REVIEW

**Comment by rodrigobrochado** [comment]
Also, does a parser PR need to update the [DATA_FORMATS_INPUT](https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_INPUT.md) file?

File: plugins/parsers/avro/parser.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+		} else {
+			log.Printf("I! AvroParser: field %s null", field)
+		}
+	}
+
+	if len(fields) == 0 {
+		log.Printf("I! AvroParser: number of fields is 0, skipping metric")
+		return nil, fmt.Errorf("Number of fields is 0, unable to create metric!")
+	} else {
+		m, err := metric.New(p.Measurement, tags, fields, metricTime)
```
[the remark this answers is not in the document]
Comment (emanuele-falzone) [reply plugins/parsers/avro/parser.go:136]: I think it is a good idea, I will update the code as suggested.

File: plugins/parsers/avro/parser.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	DefaultTags     map[string]string
+	TimeFunc        func() time.Time
+}
+
+func (p *Parser) Parse(buf []byte) ([]telegraf.Metric, error) {
+	schemaRegistry := NewSchemaRegistry(p.SchemaRegistry)
+
+	schemaID := int(binary.BigEndian.Uint32(buf[1:5]))
+
+	schema, err := schemaRegistry.getSchema(schemaID)
```
[the remark this answers is not in the document]
Comment (emanuele-falzone) [reply plugins/parsers/avro/parser.go:31]: Can you explain your use case a little more? Are you reading messages from a Kafka topic? If so, how are these messages serialized?

**Comment by sjwang90** [comment]
@rodrigobrochado Yes, a parser would need to be added to the DATA_FORMATS_INPUT readme.

**Comment by sjwang90** [comment]
@emanuele-falzone @rodrigobrochado Do either of you have any other changes to this PR you'd like to make? We'll look to get someone on the influx side to do a review if this PR is ready for review.

**Comment by rodrigobrochado** [comment]
@sjwang90 I was not able to answer some of @emanuele-falzone points yet. I think it still needs some adjustments.   

I'm sorry for the delay, I was put in another sector of our company but I'm coming back soon. When back, the matter on this PR will certainly be our priority.

File: plugins/parsers/avro/schema_registry.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	schema, ok := jsonResponse["schema"]
+	if !ok {
+		log.Printf("E! SchemaRegistry: malformed response!")
+		return "", fmt.Errorf("malformed respose from schema registry")
+	}
+
+	schemaValue, ok := schema.(string)
+	if !ok {
+		log.Printf("E! SchemaRegistry: schema %s is not a string", schema)
+		return "", fmt.Errorf("malformed respose from schema registry")
```
Comment (ssoroka) [inline plugins/parsers/avro/schema_registry.go:44]: include actual type. return or log.

**Review by ssoroka — CHANGES_REQUESTED** [review]
great start, needs some updates.

**Comment by sjwang90** [comment]
@emanuele-falzone Can you please address the comments above? Let us know if you are no longer interested in working on this.

**Comment by sjwang90** [comment]
Closing this PR due to inactivity. If someone is interested on taking this work back up, please feel free to re-open the PR and take over the code.

**Comment by athornton** [comment]
I'm starting to look at this, address the review commentary, and get it rebased on current telegraf: https://github.com/lsst-sqre/telegraf/tree/features/avro

I will issue a new PR if and when I have it working.

**Comment by athornton** [comment]
New PR is at https://github.com/influxdata/telegraf/pull/11816

**Pull request closed without merging**